src/core/models.py

```diff
@@ -0,0 +1,195 @@
+import logging
+from typing import Literal, MutableMapping
+
+from ops.model import Application, Relation, Unit
+from src.literals import CHARM_USERS
+
+logger = logging.getLogger(__name__)
+
+# TODO: move to some better literals place
+SUBSTRATE = Literal["vm", "k8s"]
+CLIENT_PORT = 2181
+SECURE_CLIENT_PORT = 2182
+SERVER_PORT = 2888
+ELECTION_PORT = 3888
+
+class StateBase:
+    def __init__(self, relation: Relation | None, component: Unit | Application, substrate: SUBSTRATE):
+        self.relation = relation
+        self.component = component
+        self.substrate = substrate
+
+    @property
+    def relation_data(self) -> MutableMapping[str, str]:
+        if not self.relation:
+            return {}
+
+        return self.relation.data[self.component]
+
+    def update(self, items: dict[str, str]) -> None:
+        if not self.relation:
+            return
+
+        self.relation_data.update(items)
+
+class ZKClient(StateBase):
```

Comment:
Absolute nitpick:

We may wanna stick to 'reusable' names, I mean something we could suggest as a standard best practice for other charms. 

Using the same terminology would make charms more readable, and help to minimize the mess and divergences from the original idea :-) 

I can only think of one suggestion here: to replace `ZK` with `Charm`. (`CharmClient`, `CharmCluster`, `CharmServer`). But there may be better alternatives.